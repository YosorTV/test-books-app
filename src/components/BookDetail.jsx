import React, { Fragment, useContext, useEffect, useState } from 'react'
import { useParams } from 'react-router';
import { Context } from '../HOC';
import { Box, Typography } from '@material-ui/core';
import { StyledCardDetail } from '../styles';
import { booksBySameAuthor } from '../helpers';
import { Link } from 'react-router-dom';

export const BookDetail = () => {
  const books = useContext(Context);
  const { bookId } = useParams();
  const [bySameAuthor, setBySameAuthor] = useState(null);
  const [content, setContent] = useState(null);

  // setting detailed book information
  useEffect(() => {
    const data = books?.find(({ id }) => id === bookId);
    setContent(data);
  },[bookId, books])
  // setting list of books by same author
  useEffect(() => {
    const data = booksBySameAuthor(books, content?.author, bookId);
    setBySameAuthor(data);
  },[books, content?.author, bookId]);

  // render list of books by same author
  const listOfBooksBySameAuthor = bySameAuthor?.map(({id, title}) => 
    <Box component="ul" key={id}>
      <Box component="li">
        <Box component={Link} to={`/all-books/${id}`}> {title} </Box>
      </Box>
    </Box>)

  // render another books section
  const anotherBooksJSX = bySameAuthor && 
  <>
    <span>Another books by author:</span> 
    {listOfBooksBySameAuthor}
  </>

  return (
    <StyledCardDetail component="article">
    {content?.author && 
      <Box component="figure" className="book-info">
        <img src={content?.cover} title={`${content?.title} by - ${content?.author}`} alt="cover" className="book-info__media"/>
        <Box component="figcaption" className="book-info__content">
          <Box className="book-info__content-description">
            <Typography component="p" className="book-info__content-description__title">
                {content?.title}
              </Typography>
              <Typography component="p" className="book-info__content-description__author">
                {content?.author}
              </Typography>
          </Box>
          <Box className="book-info__content-copy">
          {content?.description && 
            <Typography component="p" className="book-info__content-copy__about">
              {content?.description}
            </Typography>}
            <Box component="div" display="flex" flexDirection="column">
              <Typography component="p" className="book-info__content-copy__isbn">
                <span>Serial number:</span> {content?.isbn}
              </Typography>
              <Typography component="div" className="book-info__content-copy__action">
                {anotherBooksJSX} 
              </Typography>
            </Box>
          </Box>
        </Box>
      </Box>
    }
    </StyledCardDetail>
  )
}
