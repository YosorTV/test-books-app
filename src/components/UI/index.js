export * from './Loader';
export * from './Book';
export * from './Layout';