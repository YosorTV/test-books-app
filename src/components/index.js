export * from './BooksList';
export * from './BookDetail';