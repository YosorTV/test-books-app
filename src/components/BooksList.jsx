// Core
import React, { useContext } from 'react'
// MUI
import { Grid } from '@material-ui/core';

import { Context } from '../HOC'
import { Book } from './UI';

export const BooksList = () => {
  const books = useContext(Context);
  const listOfBooksJSX = books?.map((book) => <Book key={book.id} {...book} />);
  return (
    <Grid container component="ul">
      {listOfBooksJSX}
    </Grid>
  )
}
