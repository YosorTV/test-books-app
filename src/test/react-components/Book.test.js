import { render, unmountComponentAtNode } from 'react-dom';
import { BrowserRouter } from 'react-router-dom';

import Enzyme, { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import { Book } from '../../components';

Enzyme.configure({ adapter: new Adapter() });

const init = (book) => {
  const wrapper = shallow(<Book {...book} />);
  const image = wrapper.find('#img');
  const title = wrapper.find('#title');
  const action = wrapper.find('#action');

  return { wrapper, image, title, action };
};

describe('Book component', () => {
  test('Book should render without crashings', () => {
      const div = document.createElement('div');
      render(
      <BrowserRouter>
          <Book />
      </BrowserRouter>, div);
      unmountComponentAtNode(div);
  })
  test('Book action text equal to "More information"', async () => {
    const { action } = init();
    expect(action.text()).toBe('More information')
  })
  test('Book snapshot', () => {
    const { wrapper } = init();
    expect(toJSON(wrapper)).toMatchSnapshot();
  })
})

