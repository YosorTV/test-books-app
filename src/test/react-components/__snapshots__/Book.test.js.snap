// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book component Book snapshot 1`] = `
<WithStyles(ForwardRef(Grid))
  component="li"
  item={true}
  lg={4}
  md={6}
  sm={12}
  xl={4}
  xs={12}
>
  <Styled(WithStyles(ForwardRef(Card)))>
    <WithStyles(ForwardRef(CardMedia))
      className="card-image"
      id="img"
      image="placeholder.jpg"
    />
    <WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(Typography))
        component="h2"
        gutterBottom={true}
        id="title"
        variant="h5"
      />
    </WithStyles(ForwardRef(CardContent))>
    <WithStyles(ForwardRef(CardActions))>
      <WithStyles(ForwardRef(Button))
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "propTypes": Object {
              "onClick": [Function],
              "replace": [Function],
              "state": [Function],
              "target": [Function],
              "to": [Function],
            },
            "render": [Function],
          }
        }
        id="action"
        to="/all-books/undefined"
      >
        More information
      </WithStyles(ForwardRef(Button))>
    </WithStyles(ForwardRef(CardActions))>
  </Styled(WithStyles(ForwardRef(Card)))>
</WithStyles(ForwardRef(Grid))>
`;
